Add pidfd_getfd syscall

moshevds opened the pull request:
Adding the `pidfd_getfd` syscall may require some discussion, but after doing some research, I think the approach used in this PR makes some sense. So, I sent this as a PR directly instead of creating an issue first.

The 3 main things that I think require some thought, are:
 - Is this even a desired feature to add to rustix?
 - How does pidfd_getfd measure up in the I/O safety concept?
 - Are the correct types used? (Especially for `targetfd`)

In this PR, I marked the function unsafe because it _can_ be used as a kind of `dup` from a RawFd to OwnedFd, when an application uses it on itself. That would still be memory safe, but violates I/O safety as I understand from the RFC.
The argument against marking this unsafe would be that using this does not impact the safety of the calling application directly, and the safety of other processes is outside of the bounds of the concepts of memory and I/O safety anyway. (Even when the target process could end up being the calling process in practice?) I'm curious what the thoughts are about this, and what safety considerations I may have missed.

For the `targetfd` arg: I think there is no other code that uses RawFd/AsRawFd for values that are not potential file descriptors of the current process. At least to me, it feels like a bit of a stretch to use it for `targetfd`. On the other hand, using `c_int` directly seems worse. Any thoughts on this?

And the flags: The syscall currently does not take any flags, but might in the future. I added an empty enum to represent this. Is this the correct approach?

sunfishcode commented:
Thanks! I didn't know about `pidfd_getfd` before. It's surprising to me that this exists. But indeed there it is :wink: .

> The 3 main things that I think require some thought, are:
> 
>     * Is this even a desired feature to add to rustix?

Yes. Rustix in general tries not to second-guess OS features (unless they're *really* problematic like `vfork`).

>     * How does pidfd_getfd measure up in the I/O safety concept?

Great question. As you say, making this safe and using `RawFd` could make this an I/O safety bypass. But, my current thinking is, it looks like `pidfd_getfd` requires `PTRACE_MODE_ATTACH_REALCREDS` permissions over the target process, and that puts it in the category of `ptrace`-like functionality, or debugger-like functionality. So from a safety point of view, this is equivalent to attaching a debugger to the target. Rust programs can already attach debuggers to themselves and corrupt themselves, even using only safe Rust code. Consequently, it's ok for `pidfd_getfd` to be a safe function too.

This interpretation of classifing features as "debugger-like" vs. not isn't formalized anywhere, and it could evolve, but so far, it's been a rule that's been fairly straightforward to apply, and it has held up to a fair amount of informal reasoning. Rust's safety as we know it today isn't defending against malicious code; it's just defending against bugs. This suggests that we should document the hazards of `pidfd_getfd`, but we don't need to make it `unsafe`.

> For the `targetfd` arg: I think there is no other code that uses RawFd/AsRawFd for values that are not potential file descriptors of the current process. At least to me, it feels like a bit of a stretch to use it for `targetfd`. On the other hand, using `c_int` directly seems worse. Any thoughts on this?

I think it's ok to use `RawFd` for this, because it is just an integer type and can hold all the needed values. But, to better communicate it in the API, what if we defined a new type alias, like `ForeignRawFd` or some name like that, which is an alias for `RawFd`, and used it here?

> And the flags: The syscall currently does not take any flags, but might in the future. I added an empty enum to represent this. Is this the correct approach?

Yes, this looks good.

sunfishcode commented on src/process/mod.rs:
```diff
@@ -14,6 +14,8 @@ mod kill;
 #[cfg(target_os = "linux")]
 mod pidfd;
+#[cfg(linux_raw)]
+mod pidfd_getfd;
```
Rustix aims to avoid having functions available on `linux_raw` and not `libc`. `pidfd_getfd` can be supported in the `libc` backend using `libc::syscall`.

moshevds commented:
Thank you for the in-depth response. I have changed the PR as suggested. I decided to added the ForeignRawFd alias, I think it makes sense. I also changed the argument to use the alias directly, without being generic on `AsRawFd` (or a new `AsForeignRawFd` trait). I think removing the generic is better because the previous version enabled things like
```rust
pidfd_getfdpid(fd, File::open("foo.txt")?, flags)
```
which seems very wrong to me, and I cannot think of other use-cases where the generic actually makes sense for `pidfd_getfd`.

Maybe the alias should go to another place entirely? In that case I'm unsure whether the alias is really worth it, so if you/others think just using RawFd (without the generic) is better, then I can see that point as well.

sunfishcode commented on src/process/pidfd_getfd.rs:
```diff
@@ -0,0 +1,42 @@
+/// to the calling process from a value of a type that implements AsRawFd, and the target process
+/// must not drop that value until after the calling process has returned from `pidfd_getfd`.
+///
+/// When `pidfd_getfd` is used to debug the target, or the target is not a rust aplication, or
```
The word "Rust" should be capitalized within this comment.

sunfishcode commented on src/process/pidfd_getfd.rs:
```diff
@@ -0,0 +1,42 @@
+use crate::{backend, io};
+use backend::fd::{AsFd, RawFd};
+
+/// Raw file descriptor in another process.
```
It'd be good to add a note here that users should take extra care to avoid using host file descriptors here, since they use the same underlying type.

sunfishcode commented on src/process/pidfd_getfd.rs:
```diff
@@ -0,0 +1,42 @@
+/// # References
+///  - [Linux]
+///
+/// # Safety
```
Strictly speaking, this is no longer a `Safety` comment, because the function is not `unsafe`. Could you change this to say `# Warning`, which rustix uses in a few other places for similar things?

sunfishcode reviewed (commented):
Overall looks good, just a few comments about comments:

moshevds commented:
I added the suggestions, and also fixed a typo ("resuerved") that I noticed

sunfishcode commented:
Looks good!
